File: metodosOrdenacao/RUST/radix/radix.rs
const IMPRIMIR_LISTA: bool = false;

fn adquirir_maior_valor(lista: &[i32]) -> i32 {
    let mut maior = lista[0];
    for &num in lista.iter() {
        if num > maior {
            maior = num;
        }
    }
    maior
}

fn counting_sort_simplificado(lista: &[i32], lista_aux: &mut Vec<i32>, exp: i32) -> Vec<i32> {
    let mut count = vec![0i32; 10];

    for &num in lista.iter() {
        count[((num / exp) % 10) as usize] += 1;
    }

    for i in 1..10 {
        count[i] += count[i - 1];
    }

    for i in (0..lista.len()).rev() {
        let digito = ((lista[i] / exp) % 10) as usize;
        lista_aux[count[digito] as usize - 1] = lista[i];
        count[digito] -= 1;
    }

    lista_aux.to_vec()
}

// Radix Sort LSD
pub fn radix_sort(lista: &[i32]) -> Vec<i32> {
    let mut lista_atual: Vec<i32> = lista.to_vec();
    let mut lista_auxiliar: Vec<i32> = vec![0; lista.len()];

    let maior_valor = adquirir_maior_valor(lista);
    let mut expo: i32 = 1;

    if IMPRIMIR_LISTA {
        println!("Lista original: {:?}", lista);
    }

    while maior_valor / expo > 0 {
        lista_atual = counting_sort_simplificado(&lista_atual, &mut lista_auxiliar, expo);
        expo *= 10;
    }

    if IMPRIMIR_LISTA {
        println!("Lista ordenada: {:?}", lista_atual);
    }

    lista_atual
}

File: metodosOrdenacao/GO/radix/radix.go
package radix

const (
	IMPRIMIR_LISTA = false
)

func adquirir_maior_valor(lista []int) int {
	maior := lista[0]
	for i := 1; i < len(lista); i++ {
		if lista[i] > maior {
			maior = lista[i]
		}
	}

	return maior
}

func counting_sort_simplificado(lista []int, lista_aux []int, exp int) {
	count := make([]int, 10)
	tam := len(lista)
	for i := 0; i < tam; i++ {
		count[(lista[i] / exp) % 10]++
	}

	for i := 1; i < 10; i++ {
		count[i] += count[i-1]
	}

	for i := tam - 1; i >= 0; i-- {
		digito := (lista[i] / exp) % 10
		lista_aux[count[digito]-1] = lista[i]
		count[digito]--
	}

	copy(lista, lista_aux)
}

// Radix Sort LSD
func RadixSort(lista []int) {
	lista_aux := make([]int, len(lista))

	maior_valor := adquirir_maior_valor(lista)
	exp := 1

	if (IMPRIMIR_LISTA) {
		println("Lista antes da ordenação:")
		for v := range lista {
			print(lista[v], ", ")
		}
	}

	for exp = 1; maior_valor/exp > 0; exp *= 10 {
		counting_sort_simplificado(lista, lista_aux, exp)
	}

	if (IMPRIMIR_LISTA) {
		println("\nLista após a ordenação:")
		for v := range lista {
			print(lista[v], ", ")
		}
		println()
	}
}